File: Assets/WeatherVR/Scripts/Core/MapScale.cs
using UnityEngine;

namespace WeatherVR.Core
{
    /// <summary>
    /// Every conversion between real-world units and the map's local space, in one
    /// place.
    ///
    /// This is a plain struct rather than methods on <see cref="AppConfig"/> for two
    /// reasons. It can be constructed and tested outside the Unity runtime, unlike a
    /// ScriptableObject. And it makes the unit convention explicit at every call
    /// site, which matters because the failure mode here is silent: the map root
    /// carries a scale of <c>MapSizeMeters</c>, so anything under it that computes a
    /// Y coordinate in VR metres instead of in normalised map units is wrong by
    /// exactly that factor and merely looks a bit dramatic. That bug shipped once in
    /// this project — the terrain mesh built relief in metres while the cloud box and
    /// the lightning built theirs in map units, so bolts did not reach the ground.
    ///
    /// Local map space: X and Z span [-0.5, 0.5], Y uses the same units, and the map
    /// root's scale turns all three into VR metres.
    /// </summary>
    public readonly struct MapScale
    {
        /// <summary>Edge length of the map in VR metres.</summary>
        public readonly float MapSizeMeters;

        /// <summary>Edge length of the region on the ground, in real metres.</summary>
        public readonly float RegionSpanMeters;

        /// <summary>Altitude exaggeration relative to horizontal. 1 is true scale.</summary>
        public readonly float VerticalExaggeration;

        /// <summary>Extra exaggeration applied to terrain relief only.</summary>
        public readonly float TerrainReliefExaggeration;

        /// <summary>Altitude in metres that sits at the base of the rendered column.</summary>
        public readonly float AtmosphereFloorMeters;

        /// <summary>Extra exaggeration applied to building height only.</summary>
        public readonly float BuildingHeightExaggeration;

        /// <param name="buildingHeightExaggeration">
        /// Optional, defaulting to true scale, so the five-argument construction used by
        /// <c>tools/verify_logic/Verify.cs</c> and any other out-of-runtime caller keeps
        /// compiling and keeps meaning "no building exaggeration".
        /// </param>
        public MapScale(float mapSizeMeters, float regionSpanMeters, float verticalExaggeration,
                        float terrainReliefExaggeration, float atmosphereFloorMeters,
                        float buildingHeightExaggeration = 1f)
        {
            MapSizeMeters = Mathf.Max(mapSizeMeters, 1e-4f);
            RegionSpanMeters = Mathf.Max(regionSpanMeters, 1f);
            VerticalExaggeration = verticalExaggeration;
            TerrainReliefExaggeration = terrainReliefExaggeration;
            AtmosphereFloorMeters = atmosphereFloorMeters;
            BuildingHeightExaggeration = Mathf.Max(buildingHeightExaggeration, 0f);
        }

        /// <summary>VR metres per real-world metre, horizontally.</summary>
        public float Horizontal => MapSizeMeters / RegionSpanMeters;

        /// <summary>VR metres per real-world metre, vertically.</summary>
        public float Vertical => Horizontal * VerticalExaggeration;

        /// <summary>Real-world altitude in metres to VR metres above the map plane.</summary>
        public float AltitudeToVr(float altitudeMeters)
            => (altitudeMeters - AtmosphereFloorMeters) * Vertical;

        /// <summary>VR metres to normalised map units.</summary>
        public float MetersToMapUnits(float meters) => meters / MapSizeMeters;

        /// <summary>Real-world altitude in metres to map-local Y.</summary>
        public float AltitudeToMapUnits(float altitudeMeters)
            => MetersToMapUnits(AltitudeToVr(altitudeMeters));

        /// <summary>
        /// Terrain elevation in metres to map-local Y. Sea level and below flattens to
        /// zero: the sea surface is flat, and bathymetry rides in vertex colour instead.
        /// </summary>
        public float TerrainElevationToMapUnits(float elevationMeters)
            => MetersToMapUnits(Mathf.Max(elevationMeters, 0f) * Vertical * TerrainReliefExaggeration);

        /// <summary>
        /// A building's real height in metres to map-local Y, added on top of its
        /// footprint's terrain base. Deliberately uses <see cref="Horizontal"/>, not
        /// <see cref="Vertical"/>: buildings are visible at true scale in a way terrain
        /// relief and cloud altitude are not, so they do not need the atmosphere's
        /// exaggeration and would be distorted by it.
        ///
        /// <see cref="BuildingHeightExaggeration"/> is a separate, explicit multiplier on
        /// top: at 1 this is true scale (a 180 m tower on a 5 km/3 m map is 10.8 cm),
        /// above 1 the skyline is deliberately overstated. The ceiling on it is the cloud
        /// base, not taste -- exaggerate far enough and a tower punches through the deck.
        /// </summary>
        public float BuildingHeightToMapUnits(float heightMeters)
            => MetersToMapUnits(Mathf.Max(heightMeters, 0f) * Horizontal * BuildingHeightExaggeration);

        /// <summary>Denominator of the map's representative fraction, e.g. 25 000 for 1:25 000.</summary>
        public float RepresentativeFraction => 1f / Horizontal;

        public override string ToString() =>
            $"MapScale[{MapSizeMeters:F1} m map, 1:{RepresentativeFraction:N0}, " +
            $"altitude x{VerticalExaggeration:F1}, relief x{TerrainReliefExaggeration:F1}, " +
            $"buildings x{BuildingHeightExaggeration:F1}]";
    }
}

File: spatial/core/src/main/kotlin/com/weathervr/core/MapScale.kt
package com.weathervr.core

import kotlin.math.max

/**
 * Every conversion between real-world units and map-local units, in one place.
 *
 * **Read this before touching anything that positions content on the map.** Local map
 * space is a normalised unit square: X and Z run [-0.5, 0.5], Y uses the *same* units,
 * and the map root's scale ([mapSizeMeters]) turns all three into real metres.
 *
 * Any value computed in metres must therefore be divided by [mapSizeMeters] before
 * being used as a local coordinate. This shipped wrong once in the Unity version — the
 * terrain mesh built relief in metres while the cloud box and the lightning built
 * theirs in map units, so terrain came out exactly 2× too tall and bolts stopped short
 * of the ground. This class exists specifically so that cannot be got wrong silently;
 * use it for every conversion rather than multiplying by hand.
 */
class MapScale(
    mapSizeMeters: Float,
    regionSpanMeters: Float,
    /**
     * Altitude scale relative to horizontal. 1 is true scale.
     *
     * Named "exaggeration" for continuity with the Unity config, but note it ships at
     * **0.4**, which compresses: a cloud at 2 km renders at 40% of its true height above
     * the map, because at true scale the atmosphere column would leave the table
     * entirely. Both this and [terrainReliefExaggeration] are tuned for a specific
     * region span and scale inversely with it — changing the span without retuning them
     * is how a 43.7 m hill once rendered as an 84 cm spike.
     */
    val verticalExaggeration: Float,
    /** Extra exaggeration applied to terrain relief only, on top of [vertical]. */
    val terrainReliefExaggeration: Float,
    /** Altitude in metres that sits at the base of the rendered column. */
    val atmosphereFloorMeters: Float,
) {
    /** Edge length of the map, in real metres. */
    val mapSizeMeters: Float = max(mapSizeMeters, 1e-4f)

    /** Edge length of the region on the ground, in real metres. */
    val regionSpanMeters: Float = max(regionSpanMeters, 1f)

    /** Map metres per real-world metre, horizontally. */
    val horizontal: Float get() = mapSizeMeters / regionSpanMeters

    /** Map metres per real-world metre, vertically. */
    val vertical: Float get() = horizontal * verticalExaggeration

    /** Real-world altitude in metres to map metres above the map plane. */
    fun altitudeToMeters(altitudeMeters: Float): Float =
        (altitudeMeters - atmosphereFloorMeters) * vertical

    /** Real metres to normalised map units. */
    fun metersToMapUnits(meters: Float): Float = meters / mapSizeMeters

    /** Real-world altitude in metres to map-local Y. */
    fun altitudeToMapUnits(altitudeMeters: Float): Float =
        metersToMapUnits(altitudeToMeters(altitudeMeters))

    /**
     * Terrain elevation in metres to map-local Y. Sea level and below flattens to
     * zero: the sea surface is flat, and bathymetry rides in vertex colour instead.
     */
    fun terrainElevationToMapUnits(elevationMeters: Float): Float =
        metersToMapUnits(max(elevationMeters, 0f) * vertical * terrainReliefExaggeration)

    /**
     * A building's real height in metres to map-local Y, added on top of its
     * footprint's terrain base. Deliberately uses [horizontal], not [vertical]:
     * buildings are already visible at true scale (a 180 m tower on a 5 km/2 m map is
     * 7 cm), unlike terrain relief or cloud altitude, which exist to be exaggerated.
     * Exaggerating buildings too would push a skyscraper through the cloud deck.
     */
    fun buildingHeightToMapUnits(heightMeters: Float): Float =
        metersToMapUnits(max(heightMeters, 0f) * horizontal)

    override fun toString(): String =
        "MapScale[${mapSizeMeters} m map / ${regionSpanMeters} m region, " +
            "1:${(regionSpanMeters / mapSizeMeters).toInt()}, vert ×$verticalExaggeration]"
}
